bvm/ballerina-runtime/src/main/java/org/ballerinalang/jvm/MapUtils.java

```diff
@@ -148,9 +151,53 @@ public static void checkIsMapOnlyOperation(BType mapType, String op) {
         switch (mapType.getTag()) {
             case TypeTags.MAP_TAG:
             case TypeTags.JSON_TAG:
+            case TypeTags.RECORD_TYPE_TAG:
                 return;
             default:
                 throw createOpNotSupportedError(mapType, op);
         }
     }
+
+    public static void checkValidFieldForRecord(MapValue<?, ?> m, String  k, String  op) {
+        BType type = m.getType();
+        switch (type.getTag()) {
+            case TypeTags.RECORD_TYPE_TAG:
+                if (!m.containsKey(k)) {
+                    return;
+                }
+                if (isRequiredField((BRecordType) type, k)) {
+                    throw createOpNotSupportedErrorForRecord(type, op);
+                }
+                return;
+            default:
+                return;
+        }
+    }
+
+    private static boolean isRequiredField(BRecordType type, String k) {
+        Map<String, BField> fields = type.getFields();
+        BField field = fields.get(k);
+        if (field != null && Flags.isFlagOn(field.flags, Flags.REQUIRED)) {
+            return true;
+        }
+        return false;
+    }
+
+    public static void validateRecord(BType type, String op) {
+        if (type.getTag() != TypeTags.RECORD_TYPE_TAG) {
+            return;
+        }
+        Map<String, BField> fields = ((BRecordType) type).getFields();
+        for (String key : fields.keySet()) {
+            if (isRequiredField((BRecordType) type, key)) {
+                throw createOpNotSupportedErrorForRecord(type, op);
+            }
+        }
+    }
+
+    private static ErrorValue createOpNotSupportedErrorForRecord(BType type, String op) {
+        return BallerinaErrors.createError(getModulePrefixedReason(MAP_LANG_LIB,
+                OPERATION_NOT_SUPPORTED_IDENTIFIER),
+                format("%s not supported on type '%s' since there are required fields.", op, type.getQualifiedName()));
```

Comment:
Is this for the use of remove methods? If so, shall we make the error message specific? Something like below:
```
failed to remove field: 'foo' is a required field in 'FooRec'
```

Reply:
Fixed